CLI maintenance, add userdata to C API

sokorototo opened the pull request:
- Overhauled `vach-cli` to clap 4 with declarative macro to define CLI.
- Removed `split` command from `vach-cli`.
- Build `vach-cli` instead of `vach` in GitHub releases.
- Added `userdata` to C API callbacks, for closure-esque usage.
- Updated dependencies

sokorototo committed:
removed split command

sokorototo committed:
...

sokorototo committed:
added userdata to c api builder callback

sokorototo committed:
use option instead

sokorototo committed:
...

sokorototo committed:
...

sokorototo committed:
...

Signed-off-by: sokorototo <nyachiengatoto@gmail.com>

sokorototo committed:
increased test coverage, thx to cargo-llvm-cov

sokorototo committed:
update cargo dependencies

sokorototo committed:
...

sokorototo committed:
update release.yml

sokorototo committed:
updated CLI to clap 4

sokorototo committed:
refined help message

sokorototo committed:
updated docs to use clap autogenerated docs

sokorototo committed:
updated package version, as this introduces a breaking change

sokorototo committed:
...

sokorototo committed:
fixed tests

sokorototo committed:
fixed CLI on multiple arguments

zeskeertwee commented:
Just gonna try out this copilot review, see what it does.

zeskeertwee commented:
That worked better than i expected, it raised some good points. How does it actually compile with the &mut self in a const fn?

sokorototo commented on crates/vach-cli/src/commands/pack.rs:
```diff
@@ -131,50 +122,39 @@ impl CommandTrait for Evaluator {
 		// Used to filter invalid inputs and excluded inputs
 		let path_filter = |path: &PathBuf| match path.canonicalize() {
-			Ok(canonical) => !excludes.contains(&canonical) && canonical.is_file(),
+			Ok(canonical) => !excludes.iter().any(|p| canonical.starts_with(p)) && canonical.is_file(),
```
This is intentional, as now you can exclude entire folders. Maybe I should use a wildcard for this, but I was too lazy to figure out any advanced logic

sokorototo commented:
- LOL, you _can_ mutate data in a `const` context. 
 - I'll patch the `flush` code, that's a great catch.

sokorototo committed:
don't unwrap in library code 💀

sokorototo commented:
> That worked better than i expected, it raised some good points. How does it actually compile with the &mut self in a const fn?

Pretty cool, how can I invoke it?

sokorototo committed:
version bump

sokorototo committed:
a bit more refinement

allow setting a manual chunk size, if threads > entries

forbid unsafe 😁

...

rust analyzer is fucking bugged

...

zeskeertwee commented:
> > That worked better than i expected, it raised some good points. How does it actually compile with the &mut self in a const fn?
> 
> Pretty cool, how can I invoke it?

You can invoke it from the reviewers section on the top, pretty neat.

sokorototo commented:
Be sure to publish a new release 🙃
